guides/unit-testing.md

```diff
@@ -0,0 +1,973 @@
+      );
+      decryptWithKeyStub.resolves(Promise.resolve(returnValue));
+      keyringController.store.updateState({ vault: MOCK_ENCRYPTION_DATA });
+      await keyringController.setLocked();
+
+      await keyringController.submitEncryptionKey(
+        'mockEncryptionKey',
+        'mockEncryptionSalt',
+      );
+
+      expect(keyringController.encryptor.decryptWithKey.calledOnce).toBe(true);
+      expect(keyringController.keyrings).toHaveLength(1);
+    });
+  })
+});
+```
+
+### Read more
+
+- ["Arrange, act, assert"](http://xp123.com/articles/3a-arrange-act-assert/) is a simplification of the four-phase test.
+- In [behavior-driven development](https://www.agilealliance.org/glossary/bdd/), this is also called ["given, when, then"](https://martinfowler.com/bliki/GivenWhenThen.html).
+
+## Keep tests isolated
+
+In order to be useful, tests must be deterministic. That is, a single test within a file must always pass each and every time, whether it is run on its own, alongside other tests in the same file, or other tests across other files.
+
+To achieve determinism, tests must be performed in a clean room. If a test makes changes to any kind of state which persists outside of itself, it must revert those changes after completion to prevent contaminating other tests.
+
+There are a few different forms that this state can take:
+
+### Module mocks
+
+Mocks for modules — objects reachable via `import` lines — are created via `jest.mock` declarations within a test file. Since modules are mocked once when a test file is read, they are global in the context of the file, so changes in one test can "bleed" over into another test:
+
+``` typescript
+// Say `functionUnderTest` uses `config` internmally
+import { config } from "some-module";
+import { functionUnderTest } from "./module-under-test";
+
+jest.mock("some-module", () => {
+  return {
+    ...jest.requireActual("some-module"),
+    config: {
+      isActive: true
+    }
+  };
+});
+
+describe("module-under-test", () => {
+  describe("functionUnderTest", () => {
+    it("does something when isActive is false", () => {
+      // Oops! Since `config` is accessible throughout this file, this line
+      // changes not only this test but every other test after this
+      config.isActive = false;
+      expect(functionUnderTest()).toDoSomething();
+    });
+
+    it("does something usually", () => {
+      // Now `config.isActive` will be `false` within `functionUnderTest`,
+      // so it may not do what you expect
+      expect(functionUnderTest()).toDoSomething();
+    })
+  });
+})
+```
+
+Setting Jest's [`resetModules`](https://jestjs.io/docs/configuration#resetmodules-boolean) option to `true` instructs Jest to recreate the module registry for each test (i.e. rerun the function that you pass to `jest.mock`), guaranteeing that state kept in modules gets reset appropriately. (This option is set in MetaMask's [module template](https://github.com/MetaMask/metamask-module-template).)
```

Comment:
Hmm. Technically we set `resetMocks` not `resetModules`, though mocks includes module mocks. The distinction seems unnecessary to explain here though. Maybe to avoid this confusion we can reference the `resetMocks` configuration option here instead?

Reply:
Actually.... does `resetMocks` include `resetModules`? I swear I've looked this up before but now I'm not sure. The docs are frustratingly ambiguous about this.

Reply:
Oh, interesting. So according to the documentation:

- `resetMocks` calls `jest.resetAllMocks()` before each test. `jest.resetAllMocks()` calls `.mockReset()` on every mocked function.
- `resetModules` calls `jest.resetModules()` before each test. `jest.resetModules()` resets the module registry.
- `restoreMocks` calls `jest.restoreAllMocks()` before each test. `jest.restoreAllMocks()` calls `.mockRestore()` on every mocked function and `.restore()` on every replaced property. It only works on mocks created via `jest.spyOn()` and properties created with `jest.replaceProperty()`.

But according to the comments in `jest.config.js`:

- `resetMocks` resets all mocks, including mocked modules, to `jest.fn()` between each test case.
- `resetModules` resets the module registry before running each individual test.
- `restoreMocks` restores all mocks created using `jest.spyOn()` to their original implementation between each test. It does not affect mocked modules.

So... I'm unsure. According to the source code, [`jest.resetAllMocks()` calls `resetAllMocks` on a ModuleMocker instance](https://github.com/jestjs/jest/blob/331ad059334ab024d0f395cb49fd220d4bebd805/packages/jest-runtime/src/index.ts#L1351), but [`jest.resetModules()` does a bunch of stuff to reset the module registry](https://github.com/jestjs/jest/blob/331ad059334ab024d0f395cb49fd220d4bebd805/packages/jest-runtime/src/index.ts#L1197). So I am inclined to think that the comment for `resetAllMocks` is lying and that it _doesn't_ actually handle modules. But I'll try it out.

Reply:
After playing around with `resetModules`, it appears that it won't protect against the use case I presented in this "Module mocks" section. I think we would need to re-`require` module between each test if we really wanted to do that. That's a bit disappointing, but maybe I will experiment some more later. In the meantime I can remove this section.

Reply:
And your suggestion to recommend the use of the configuration option and put the details in a "Read more" section is good one, I can make that change.

Reply:
Okay, I revised this section to recommend using `resetMocks` and `restoreMocks`: [45403ef](https://github.com/MetaMask/contributor-docs/pull/38/commits/45403ef859a71d02a80596b6b6b6dfd810cec0c9)